bindings/node/index.d.ts

```diff
@@ -0,0 +1,340 @@
+   * @param result If present, is the decrypted result
+   */
+  (error?: Error, result?: any): void;
+}
+
+/**
+ * Configuration options that are used by specific KMS providers during key generation, encryption, and decryption.
+ */
+export interface KMSProviders {
+  /**
+   * Configuration options for using 'aws' as your KMS provider
+   */
+  aws?: {
+    /**
+     * The access key used for the AWS KMS provider
+     */
+    accessKeyId?: string | undefined;
+
+    /**
+     * The secret access key used for the AWS KMS provider
+     */
+    secretAccessKey?: string | undefined;
+  };
+
+  /**
+   * Configuration options for using 'local' as your KMS provider
+   */
+  local?: {
+    /**
+     * The master key used to encrypt/decrypt data keys. A 96-byte long Buffer.
+     */
+    key?: Buffer | undefined;
+  };
+
+  /**
+   * Configuration options for using 'azure' as your KMS provider
+   */
+  azure?: {
+    /**
+     * The tenant ID identifies the organization for the account
+     */
+    tenantId?: string | undefined;
+
+    /**
+     * The client ID to authenticate a registered application
+     */
+    clientId?: string | undefined;
+
+    /**
+     * The client secret to authenticate a registered application
+     */
+    clientSecret?: string | undefined;
+
+    /**
+     * If present, a host with optional port. E.g. "example.com" or "example.com:443".
+     * This is optional, and only needed if customer is using a non-commercial Azure instance
+     * (e.g. a government or China account, which use different URLs).
+     * Defaults to "login.microsoftonline.com"
+     */
+    identityPlatformEndpoint?: string | undefined;
+  };
+  
+  /**
+   * Configuration options for using 'gcp' as your KMS provider
+   */
+  gcp?: {
+    /**
+     * The service account email to authenticate
+     */
+    email?: string | undefined;
+
+    /**
+     * A PKCS#8 encrypted key. This can either be a base64 string or a binary representation
+     */
+    privateKey?: string | Buffer | undefined;
+
+    /**
+     * If present, a host with optional port. E.g. "example.com" or "example.com:443".
+     * Defaults to "oauth2.googleapis.com"
+     */
+    endpoint?: string | undefined;
+  }
+}
+
+/**
+ * Optional settings to provide when creating a new `ClientEncryption` instance.
+ */
+export interface ClientEncryptionOptions {
+  /**
+   * The namespace of the key vault, used to store encryption keys
+   */
+  keyVaultNamespace: string;
+
+  /**
+   * A MongoClient used to fetch keys from a key vault. Defaults to client.
+   */
+  keyVaultClient?: MongoClient | undefined;
+
+  /**
+   * Options for specific KMS providers to use
+   */
+  kmsProviders?: KMSProviders | undefined;
+}
+
+/**
+ * Configuration options for making an AWS encryption key
+ */
+export interface AWSEncryptionKeyOptions {
+  /**
+   * The AWS region of the KMS
+   */
+  region?: string | undefined;
```

Comment:
Since I am new to Typescript, if I understand this correctly, this type definition means I could have a object of type `AWSEncryptionKeyOptions` without `region` set?

I ask this because `region` and `key` are required for an AWS key. The `endpoint` field is optional.

Reply:
Yes, your understanding is correct.

Right now the types are very "unrestrictive" as I was looking at the documentation of the library when creating that. Do you know what the best place is to grab the exact details on which parameters are required and which are optional? Can I rely on what is defined in https://github.com/mongodb/specifications/blob/master/source/client-side-encryption/client-side-encryption.rst#user-facing-api?

Reply:
Yes, the best place for the grab to the exact details is in that spec.
https://github.com/mongodb/specifications/blob/master/source/client-side-encryption/client-side-encryption.rst#kmsproviders
https://github.com/mongodb/specifications/blob/master/source/client-side-encryption/client-side-encryption.rst#masterkey